### PR TITLE
fix: don't drop single `BaseModel` tool arguments sent in wrapped form when the model uses default `extra='ignore'`

### LINKED ISSUE #5550: single-arg unwrap silently dropping wrapped payloads
### Initial Checks

- [x] I'm using the [latest version](https://github.com/pydantic/pydantic-ai/releases/latest) of Pydantic AI
- [x] I've searched for my issue in [the issue tracker](https://github.com/pydantic/pydantic-ai/issues) before opening this issue

### Description

Context
---
For a tool function with a single ``BaseModel`` parameter, pydantic-ai flattens the input schema sent to the model (so the LLM emits the model fields at the top level instead of inside a redundant wrapper). The ``_validate_single_arg`` wrap validator accepts BOTH shapes — unwrapped ``{name: ..., value: ...}`` and wrapped ``{<param>: {name: ..., value: ...}}`` — to support callers that re-validate already-validated args (e.g. Temporal's serialization round-trip, or LLMs that didn't honor the unwrap).

The validator tries the unwrapped path first and only unwraps the ``{<param>: {...}}`` envelope on ``ValidationError``:

```python
    def _validate_single_arg(value, handler, *, name):
        try:
            return {name: handler(value)}      # try unwrapped
        except ValidationError:
            if isinstance(value, dict) and list(value) == [name]:
                return {name: handler(value[name])}  # fallback: unwrap
            raise
```

Bug
---
When the input model uses pydantic's default ``extra="ignore"``, validating ``{<param>: {...}}`` against the unwrapped schema does NOT raise — the ``<param>`` key is silently dropped as an extra, and every field falls back to its default. The fallback never runs and the tool receives a defaulted model.

Reproduction
------------
Run with:

```sh
    uv run pytest /tmp/test_pydantic_ai_single_arg_bug.py -v
```
On pydantic-ai 1.96.0 the assertions fail: the tool receives ``MyInput(name='default_name', value=0)`` instead of the ``MyInput(name='actual_name', value=42)`` payload the model sent.

A possible fix is to attempt the wrapped form first when ``value`` is shaped like ``{<param>: dict}``, or to make the validator reject extras at this level (or make it configurable). The current implementation is very fragile.
"""

### Minimal, Reproducible Example

```Python
from __future__ import annotations

from pydantic import BaseModel, ConfigDict
from pydantic_ai import Agent
from pydantic_ai.messages import (
    ModelMessage,
    ModelResponse,
    TextPart,
    ToolCallPart,
)
from pydantic_ai.models.function import AgentInfo, FunctionModel


class MyInput(BaseModel):
    # Default pydantic behavior: extras are silently ignored. This is the
    # configuration that triggers the bug.
    name: str = "default_name"
    value: int = 0


class StrictInput(BaseModel):
    # With extras forbidden, the unwrapped attempt against a wrapped payload
    # raises ValidationError and the fallback path correctly unwraps.
    model_config = ConfigDict(extra="forbid")

    name: str = "default_name"
    value: int = 0


def _build_agent(tool, wrapped_args: dict) -> Agent:
    call_count = 0

    async def model_fn(messages: list[ModelMessage], info: AgentInfo) -> ModelResponse:
        nonlocal call_count
        call_count += 1
        if call_count == 1:
            # Emit the WRAPPED shape. This is what an LLM that didn't honor
            # the unwrap hint emits, and what re-validation paths (e.g.
            # Temporal serialization round-trip) feed back through the
            # validator. Either way, it should round-trip through
            # _validate_single_arg without data loss.
            return ModelResponse(
                parts=[ToolCallPart(tool_name=tool.__name__, args=wrapped_args)]
            )
        return ModelResponse(parts=[TextPart("done")])

    return Agent(FunctionModel(model_fn), tools=[tool])


def test_single_arg_default_extras_loses_wrapped_payload():
    """Demonstrates the bug: with default extras=ignore, wrapped args are silently dropped."""
    received: list[MyInput] = []

    def my_tool(input: MyInput) -> str:
        received.append(input)
        return "ok"

    agent = _build_agent(
        my_tool, wrapped_args={"input": {"name": "actual_name", "value": 42}}
    )
    agent.run_sync("go")

    assert len(received) == 1
    got = received[0]
    # Both assertions FAIL on pydantic-ai 1.96.0 — got is MyInput() with defaults.
    assert got.name == "actual_name", (
        f"name was {got.name!r}: the wrapped payload was silently dropped"
    )
    assert got.value == 42, (
        f"value was {got.value}: the wrapped payload was silently dropped"
    )


def test_single_arg_forbid_extras_unwraps_correctly():
    """Control: with extra='forbid', the fallback path runs and unwrapping works."""
    received: list[StrictInput] = []

    def strict_tool(input: StrictInput) -> str:
        received.append(input)
        return "ok"

    agent = _build_agent(
        strict_tool, wrapped_args={"input": {"name": "actual_name", "value": 42}}
    )
    agent.run_sync("go")

    assert len(received) == 1
    got = received[0]
    assert got.name == "actual_name"
    assert got.value == 42
```

### Logfire Trace

_No response_

### Python, Pydantic AI & LLM client version

- Python: 3.12.13
- Pydantic AI: 1.96.0

### PR DESCRIPTION
**Pull request by macayu17**
### Summary

Fixes #5550

A tool whose single argument is a model-like type (`BaseModel`, dataclass, `TypedDict`) has its schema flattened, so the model emits the fields at the top level. The validator accepts both that unwrapped shape *and* the already-wrapped `{argument: {...}}` shape, so that previously-validated args stay valid when re-validated (e.g. across a Temporal activity round-trip).

The bug: the validator tried the unwrapped shape first and only unwrapped `{argument: {...}}` on `ValidationError`. Under Pydantic's default `extra='ignore'`, validating the wrapped shape against the unwrapped schema does **not** raise — `argument` is dropped as an extra and every field falls back to its default — so the real payload was silently lost.

### Changes

- Disambiguate `{argument: ...}` by whether `argument` is a key the model actually **accepts** (field name *or* alias), read from the model's validation JSON schema so Pydantic resolves aliases (incl. `AliasChoices`) for us.
- If `argument` is not an accepted key, `{argument: ...}` can only be a wrapper envelope → unwrap it.
- If it is a real field/alias, validate as-is and fall back to unwrapping only on `ValidationError` (the round-trip of such a model).
- Regression tests: wrapped payload with defaults (the reported case), same-named-field round-trip, aliased field with a default, and `AliasChoices`.

### Known limitations

- **Permissive name collision (irreducible):** if the model's field shares the parameter's name *and* accepts both the payload and the wrapped payload (e.g. typed `Any`), `{name: {name: X}}` validates as either reading. We prefer the unwrapped one, so re-validation isn't idempotent for that rare case. This matches `main`'s behavior; eliminating it would require unwrapping at the re-validation boundary rather than in the validator.
- **`populate_by_name` + alias:** a bare `{field_name: value}` (using the canonical name rather than the alias) for such a model is treated as a wrapper and raises. No LLM or Temporal round-trip emits this shape — they use the alias — so it only affects hand-written `validate_python` calls.

### Checklist

- [x] Any **AI generated code** has been reviewed line-by-line by the human PR author, who stands by it.
- [x] No **breaking changes** in accordance with the [version policy](https://github.com/pydantic/pydantic-ai/blob/main/docs/version-policy.md).
- [x] **PR title** is fit for the [release changelog](https://github.com/pydantic/pydantic-ai/releases).

### CODE REVIEW

**Comment by macayu17** [comment]
No worries, I missed that #5553 was already open before creating mine. My branch had a regression test for the default-value case and also covered the edge case where the model itself has a top-level field with the same name as the tool parameter. Happy to leave it here if that helps with the existing PR.

**Comment by DouweM** [comment]
@macayu17 Sorry about the auto-close, I'll check this out in a bit.

File: pydantic_ai_slim/pydantic_ai/_function_schema.py [hunk 1]
```diff
@@ -338,20 +349,43 @@ def _build_schema(
...
 
 
+def _annotation_accepts_key(annotation: Any, name: str) -> bool:
+    """Whether a model-like `annotation` accepts `name` as a top-level validation key.
+
+    Used by `_build_schema` to tell a wrapper envelope `{name: ...}` (which we unwrap) apart from
+    genuine unwrapped input for a model with a field — or field alias — called `name` (which we don't).
+    Reading the keys from the validation JSON schema lets Pydantic resolve aliases for us.
+    """
+    properties = TypeAdapter(annotation).json_schema(mode='validation').get('properties', {})
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/_function_schema.py]: This is reasonably expensive. Don't we already generate the type adapter somewhere? Could/should we reuse that?
